least squares and gaussian elimination

gravilet opened the pull request:
still not realized some notes :
- divisor = 0
- matrix transformations into standalone function
-- openblast

zingorn committed:
least squares and gaussian elimination - first realization

zingorn committed:
cleanup + coding style fixing

gravilet committed:
coding style fixing
add some smoke tests
fix gaussian elimination logic

alexey-sidnev commented on include/gaussian_elimination.h:
```diff
@@ -0,0 +1,76 @@
+            }
+        }
+        if (!flag) {
+            return 2;       //
```
Add `enum class` for error codes.

gravilet replied:
done. в отдельном файле для всего нейспейса

alexey-sidnev commented on include/gaussian_elimination.h:
```diff
@@ -0,0 +1,76 @@
+     * поэтому - храним в отдельном set-е набор индексов, которые уже отнормировали,
+     * ищем в столбце первый ненулевой элемент не из этого set-а и нормируем по нему
+     */
+    std::set<size_t> normalized;
```
Replace it with `vector`.

gravilet replied:
done. + выбирается макс. элемент для лучшей точности => индексные массивы, чтобы не запутаться в порядке нормирования элементов

alexey-sidnev commented on include/least_squares.h:
```diff
@@ -0,0 +1,80 @@
+
+    int result_code = sole_solver(&sole_matrix, &sole_constant_terms);
+    if (result_code == 0) {
+        *result = sole_constant_terms;
```
Use `std::move()`

gravilet replied:
done

alexey-sidnev commented on test/least_squares_test.cpp:
```diff
@@ -0,0 +1,61 @@
+    }
+}
+
+TEST(Smoke_test, manually_precalculated_examples___ax_bxx_cxxx) {
```
Use `SmokeTest`

gravilet replied:
done

gravilet committed:
fix gaussian elimination logic
add enum for error codes
replace 'size_t' by another types

DmitriySidnev committed:
Fix CMakeLists for gtest

alexey-sidnev committed:
Update CMake rules for cpplint

1. Update CMake cpplint rules
2. Add `build/include` filter to cpplint
3. Remove commented lines
4. Fix minor linting error

gravilet committed:
some tuning

gravilet committed:
some cleanup
